vhost-device-vsock/src/main.rs

```diff
@@ -266,20 +269,32 @@ pub(crate) fn start_backend_server(
 
 pub(crate) fn start_backend_servers(configs: &[VsockConfig]) -> Result<(), BackendError> {
     let cid_map: Arc<RwLock<CidMap>> = Arc::new(RwLock::new(HashMap::new()));
-    let mut handles = Vec::new();
+    let mut handles = HashMap::new();
+    let (senders, receiver) = std::sync::mpsc::channel();
 
-    for c in configs.iter() {
+    for (thread_id, c) in configs.iter().enumerate() {
         let config = c.clone();
         let cid_map = cid_map.clone();
+        let sender = senders.clone();
+        let name = format!("vhu-vsock-cid-{}", c.get_guest_cid());
         let handle = thread::Builder::new()
-            .name(format!("vhu-vsock-cid-{}", c.get_guest_cid()))
-            .spawn(move || start_backend_server(config, cid_map))
+            .name(name.clone())
+            .spawn(move || {
+                let result =
+                    std::panic::catch_unwind(move || start_backend_server(config, cid_map));
+
+                // Notify the main thread that we are done.
+                sender.send(thread_id).unwrap();
+
+                result.map_err(|e| BackendError::ThreadPanic(name, e))?
+            })
             .unwrap();
-        handles.push(handle);
+        handles.insert(thread_id, handle);
     }
 
-    for handle in handles {
-        handle.join().unwrap()?;
+    while !handles.is_empty() {
+        let thread_id = receiver.recv().unwrap();
+        handles.remove(&thread_id).unwrap().join().unwrap()?;
```

Comment:
Could we collect all handle return values? In case there is more than one panic. Just a thought, not a suggestion.

Reply:
Yep, I also thought about that, but how we should handle them?
Print an error here and make this function void without returning anything to the main? (makes sense to me)

Reply:
Hm, something like..

Receive from `receiver` `handles.len()` times, then collect join errors:

`handles.into_iter().filter_map(|v| if let Err(err) = v { Some(err) } else { None }).collect::<Vec<_>>()`

Reply:
Sorry, I didn't get :-(
We can't wait for all the threads, otherwise we have the same problem we had before.

When we have an error, we have to print it right away or exit, because otherwise the error would only be printed when all the threads exit. The ones without errors, they never exit.

Reply:
Ah that's right! Nevermind that then.